### PR TITLE
Fix bug in make_cannonical, make Child.ref a macro

### PR DESCRIPTION
**Pull request by FelonEkonom**
Related Jira tickets:
 - https://membraneframework.atlassian.net/browse/MC-169
 - https://membraneframework.atlassian.net/browse/MC-170

### CODE REVIEW

File: lib/membrane/child.ex [hunk 1]
```diff
@@ -21,22 +20,68 @@ defmodule Membrane.Child do
...
+    end
+  end
+
+  defp ref_in_a_match(name) do
+    quote do
+      unquote(name)
+    end
+  end
+
+  defp ref_in_a_match(name, group: group) do
```
Comment (varsill) [inline lib/membrane/child.ex:43]: Right now, when the user passes an improper list of options to the `Child.ref` in `match` context, an error saying that the "`ref_in_a_match` function couldn't be matched" is raised. How about adding the following function clause:
```
 defp ref_in_a_match(_name, improper_options) do
    raise "Improper options: #{inspect(improper_options)}. The options must be in form of [group: group]."
  end
```

Comment (varsill) [reply lib/membrane/child.ex:43]: Perhaps I am running into the future to much, but I also wonder if we should allow the user to match on the whole options list, that is - to fetch the `opts` from the following match case: `Child.ref(child_name, opts) = child`.

File: lib/membrane/child.ex [hunk 2]
```diff
@@ -21,22 +20,68 @@ defmodule Membrane.Child do
...
+  defmacro ref(name, options) do
+    case __CALLER__.context do
+      :match -> ref_in_a_match(name, options)
+      _not_match -> ref_outside_a_match(name, options)
+    end
+  end
+
+  defp ref_in_a_match(name) do
+    quote do
+      unquote(name)
```
Comment (varsill) [inline lib/membrane/child.ex:39]: why would we want to allow for such a match: `Child.ref({Membrane.Child, _, _}) = {Membrane.Child, :group, :name}`?

**Review by varsill — CHANGES_REQUESTED** [review]

File: test/membrane/integration/accepted_format_test.exs [hunk 3]
```diff
@@ -89,6 +89,7 @@ defmodule Membrane.StreamFormatTest do
   defp assert_down(module) do
     assert_receive({:my_pid, ^module, pid})
     Process.monitor(pid)
-    assert_receive({:DOWN, _ref, :process, ^pid, {%Membrane.StreamFormatError{}, _stacktrace}})
+    assert_receive({:DOWN, _ref, :process, ^pid, reason})
+    assert match?({%Membrane.StreamFormatError{}, _stacktrace}, reason) or reason == :noproc
```
[the remark this answers is not in the document]
Comment (FelonEkonom) [reply test/membrane/integration/accepted_format_test.exs:93]: When `pid` dies before `Process.monitor`

File: lib/membrane/element/base.ex [hunk 4]
```diff
@@ -243,13 +243,20 @@ defmodule Membrane.Element.Base do
 
   Options:
     - `:bring_pad?` - if true (default) requires and aliases `Membrane.Pad`
+    - `:bring_child?` - if true (default) requires and aliases `Membrane.Child`
```
[the remark this answers is not in the document]
Comment (FelonEkonom) [reply lib/membrane/element/base.ex:246]: Element still can use it to match on it's own name

File: lib/membrane/child.ex (hunk 2 above)
Comment (FelonEkonom) [reply lib/membrane/child.ex:39]: I don't, but I cannot check in inside a match :shit:

**Review by varsill — APPROVED** [review]

File: test/membrane/integration/accepted_format_test.exs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mat-hek) [reply test/membrane/integration/accepted_format_test.exs:93]: This approach smells, if the reason is `:noproc` then the element could have died for any reason or even never started. The appropriate way would be to monitor the process when we're sure it should be still alive or use start_link

**Review by mat-hek — APPROVED** [review]

**Review by mat-hek — APPROVED** [review]

**Pull request merged**